src/main.cc

```diff
@@ -122,110 +161,95 @@ int main(int argc, char *argv[]) {
             pch ? pch + 1 : "",
             run_bfs ? "bfs" : "",
             run_curiosity ? curiosity_map[curiosity_policy].c_str() : "",
-            use_prefix_perm_map ? "with_prefix_perm_map" : 
+            use_prefix_perm_map ? "with_prefix_perm_map" :
                 (use_captured_sym_map ? "with_captured_symmetry_map" : "no_pmap"),
             meta ? "minor" : "no_minor",
             ablation ? ((ablation == 1) ? "support" : "lookahead") : "none",
-            max_num_nodes, c, verbosity, freq);
-    snprintf(log_fname, BUFSZ, "%s.txt", froot);
-    snprintf(opt_fname, BUFSZ, "%s-opt.txt", froot);
+            max_num_nodes, c, verbstr, freq);
+    _snprintf(log_fname, BUFSZ, "%s.txt", froot);
+    _snprintf(opt_fname, BUFSZ, "%s-opt.txt", froot);
 
-    if (verbosity >= 1000) {
-        printf("\n%d rules %d samples\n\n", nrules, nsamples);
-        rule_print_all(rules, nrules, nsamples);
+    int ret = 0;
 
-        printf("\nLabels (%d) for %d samples\n\n", nlabels, nsamples);
-        rule_print_all(labels, nlabels, nsamples);
-    }
+    if(run_corels_begin(c, &verbstr[0], curiosity_policy, map_type, ablation, calculate_size,
+                        nrules, nlabels, nsamples, rules, labels, meta, freq, &log_fname[0]) == 0)
+    {
+        while(run_corels_loop(max_num_nodes) == 0) { }
 
-    if (verbosity > 1)
-        logger = new Logger(c, nrules, verbosity, log_fname, freq);
-    else
-        logger = new NullLogger();
-    double init = timestamp();
-    char run_type[BUFSZ];
-    Queue* q;
-    strcpy(run_type, "LEARNING RULE LIST via ");
-    char const *type = "node";
-    if (curiosity_policy == 1) {
-        strcat(run_type, "CURIOUS");
-        q = new Queue(curious_cmp, run_type);
-        type = "curious";
-    } else if (curiosity_policy == 2) {
-        strcat(run_type, "LOWER BOUND");
-        q = new Queue(lb_cmp, run_type);
-    } else if (curiosity_policy == 3) {
-        strcat(run_type, "OBJECTIVE");
-        q = new Queue(objective_cmp, run_type);
-    } else if (curiosity_policy == 4) {
-        strcat(run_type, "DFS");
-        q = new Queue(dfs_cmp, run_type);
-    } else {
-        strcat(run_type, "BFS");
-        q = new Queue(base_cmp, run_type);
-    }
-
-    PermutationMap* p;
-    if (use_prefix_perm_map) {
-        strcat(run_type, " Prefix Map\n");
-        PrefixPermutationMap* prefix_pmap = new PrefixPermutationMap;
-        p = (PermutationMap*) prefix_pmap;
-    } else if (use_captured_sym_map) {
-        strcat(run_type, " Captured Symmetry Map\n");
-        CapturedPermutationMap* cap_pmap = new CapturedPermutationMap;
-        p = (PermutationMap*) cap_pmap;
+        int* rulelist = NULL;
+        int rulelist_size = 0;
+        int* classes = NULL;
+
+        run_corels_end(&rulelist, &rulelist_size, &classes, 0, latex_out, rules, labels, &opt_fname[0]);
+
+        if(rulelist)
+            free(rulelist);
```

Comment:
Are you mallocing these? It's usually good form to put the allocation near the definition so it's clear that they're on the heap.

Reply:
I malloc them in run_corels_end because I can't know their size before running bbound_end within run_corels_end. I could malloc them in main.cc with as a fixed-width buffer, but then there'd be an upper bound on the length of the rulelist that could be outputted.

Reply:
Why don't you make them vectors so you can add to them dynamically? This would align with the types we use for `get_prefix_and_predictions` more closely anyways